### ISSUE TITLE
Unable to Configure an Apple //e with 64K only

### ISSUE THREAD

**Issue by paulhlee1967** [AppleWin/AppleWin#1341]
Trying to configure Applewin as an Apple //e (both enhanced and unenhanced) with only 64K of RAM.  Tried using the command line switch -r 0 and still it shows 128K,

Am I missing something or is this a real bug?

**Comment by tomcw** [comment]
It's a feature that hasn't been implemented (so //e's always have 128K minimum).
I can take this as a request and use this ticket to track progress.

**Comment by tomcw** [comment]
Some quick test code:
300:8D 0 C0 8D 5 C0 A2 11 8E 20 4 E8 8E 21 4 E8 8E 22 4 E8 8E 23 4 8D 4 C0 60

```
0300:8D 00 C0 q           STA KBD        C000  ; 80STOREOFF
0303:8D 05 C0             STA WRCARDRAM  C005
0306:A2 11                LDX #$11
0308:8E 20 04             STX $0420      0420
030B:E8                   INX                
030C:8E 21 04             STX $0421      0421
030F:E8                   INX                
0310:8E 22 04             STX $0422      0422
0313:E8                   INX                
0314:8E 23 04             STX $0423      0423
0317:8D 04 C0             STA WRMAINRAM  C004

```

After unplugging my "Extended 80 Column" card from my //e...

With 80STOREOFF (300:00), then no chars are written to the screen.
With 80STOREON (300:01), then "QRST" are written to the screen.

So far, all the same, as if the 80 col card were still plugged in.

**Comment by tomcw** [comment]
Enable 80COL (C00D:0), then the same char is fetched for odd/even columns...

Well not quite, for the first 8 rows it's prefetching $FF(?) (checkerboard char), then doubling up every char across the row, and then the 79th char appears further down the screen (start or row 9).

![DSC_1963](https://github.com/user-attachments/assets/bc0ce801-be0c-49e1-a5b9-ea2113043ce1)

![DSC_1964](https://github.com/user-attachments/assets/54b40991-7063-4166-a00a-fef61a6579bc)

**Comment by tomcw** [comment]
I changed the code so that the first two writes are to $320 and $321 (instead of $420 and $421):
- result: nothing gets written to $320-321.
- so writes to aux memory are lost.
- unless 80STOREON && PAGE2OFF (and the writes are to $400-7FF or $2000-3FFF).

**Comment by tomcw** [comment]
Some quick test code:
180: 8D 3 C0 AD 0 3 8D 20 3 AD 2 3 8D 21 3 8D 2 C0 60

```
0180:8D 03 C0             STA RDCARDRAM  C003
0183:AD 00 03             LDA $0300      0300
0186:8D 20 03             STA $0320      0320
0189:AD 02 03             LDA $0302      0302
018C:8D 21 03             STA $0321      0321
018F:8D 02 C0             STA RDMAINRAM  C002
```

Still with my "Extended 80 Column" card unlugged from my //e...

Beforehand:
300: 8D -- C0
320: 00 00

After running the code: (FC58G N 180G)
320: A0 A0

So it's just reading the floating bus when aux memory is enabled ($0200-BFFF).

**Comment by tomcw** [comment]
With the card unplugged, boot ProDOS:
BF98: A0 = IIe, 64K, no 80-col card (see #1321)

**Comment by tomcw** [comment]
I tried MAME 0.267 (apple2ee, -aux ""), and that doesn't show this odd $FF aux video prefetch:

![mame-0267](https://github.com/user-attachments/assets/96f110cd-641f-4010-bb0e-9a14543f2450)

**Comment by tomcw** [comment]
Sather (UTAIIe, 5-38):
> if an attempt to read auxiliary card RAM is made and no card is installed, the MPU will read the last value driven out of motherboard memory by the video scanner.

So with 80COL enabled, the video generator is just reading the floating bus, hence the repeat video characters.
NB. The video generator order is aux value (even column) then main value (odd column).

For row 0,
- col 0 aux - 0x7F or 0xFF from floating bus
- col 0 main
- col 1 aux - repeat value from "col 0 main" from floating bus
- ...
- col 39 aux - repeat value from "col 38 main" from floating bus
- col 39 main

And so this is why the first 80COL video character in col 0/aux isn't repeated in col 0/main, rather the repeat starts from col 1/aux; or in 80 column counting (ie. columns 0-79): repeats are col 1&2, col 3&4, ..., col 77&78.

**Comment by tomcw** [comment]
Yeah, so I assume my IIe powers-up with 0x7F or 0xFF in $47F, and then if I change that value it'll be displayed at (row 0) col 0 aux.
I'll check this tomorrow.

**Comment by univta0001** [comment]
Hi, there are two configurations for Apple //e with 64K

1. By removing the auxiliary card from the auxiliary slot. Therefore 80-column will no longer works.
2. Replace the extended 80 column card (64 KB) with a standard 80 column card with 1 KB memory. This will allow the 80-column to work and reduce the apple //e to 64KB. In this mode, double hires and double lores do not work.

Option 2 is easier to implement by limit the aux memory address access up to 0xc00.

The details on how to detect the Aux card being removed or using 1K Aux Card is documented at https://prodos8.com/docs/technote/misc/02/

**Comment by tomcw** [comment]
NB. Once implemented, then I need to check that VidHD's SHR works (aux= empty / 80col / extended 80 col).
EG. Using TR, press CTRL+@ to run the SHR slideshow.

**Comment by tomcw** [comment]
And check... aux: empty or std80

```
C050 N C057 N C05E		; TEXTOFF, HIRESON, DHIRESON: just HIRES magenta & green
C00D:0				; 80COL: no affect (ie. should not switch to DHIRES, like ext80 does)

C050 N C05E			; TEXTOFF, DHIRESON: remains in GR
				; (not the funky GR HIRES magenta & green mode, like ext80 does)
C00D:0				; 80COL: no affect (ie. should not switch to DGR, like ext80 does)
```

**Comment by tomcw** [comment]
**Note to self:**
My Cirtech 80-col (1KiB) card, ie. "std80" which is apparently faulty, as ProDOS, $BF98 = 0xA0 (ie. //e, 64K, no 80-col).

I tested with this card, and Test 1 displays 7M GR (then C00D:0 switches to DGR), and Test 2 displays 7M HGR  (then C00D:0 switches to DHGR).

I also tried the (Nikrom?) Master Diagnostics //e, which reports //e, 64K, no 80-col - maybe this is just using the ProDOS $BF98 value?

**Comment by xotmatrix** [comment]
Is it just that card? I've become skeptical of some of the 80-column detection schemes ProDOS uses. I have an Apple 1KB 80-column card but all my stuff is in storage, so I can't test.

**Comment by tomcw** [comment]
I suspect it's the RAM chips on the card (2x 2114, which are 1024 x 4 bit static RAM chips).
I need to try lifting them, and probably replacing them.

**Comment by tomcw** [comment]
The commit (16b2cf3) is "good enough" to convince ProDOS(*), Total Replay, Master Diagnostics //e that it's just a 1KiB 80-col card, but a2audit gets confused (and doesn't support this 1K 80-col card) and the output should currently look like this: https://github.com/zellyn/a2audit/issues/16.

(*) **BF98: A2**  ; //e, only 64K, 80-Col
Whereas for a 128K //e, **BF98: B2**, and for a //e with empty aux slot: **BF98: A0**

Bottom line is that AppleWin's 1KiB 80-col card emulation can be improved.

**Comment by Michaelangel007** [comment]
@tomcw Here is the summary of tasks from that PR.  Did you want me to make a new issues for these or handle them here?

* [ ] Update Help documents, Configuration Settings, mentioning that one can use `Ctrl-`C` to copy the benchmark.
* [ ] Update the Benchmark output to include AppleWin's version in the output first thing.
* [ ] Create a native 64-bit Configuration
* [ ] Update the Debugger's `CmdMOTD()` to display the platform: i.e. `x86` or `x64`

**Comment by tomcw** [comment]
>Did you want me to make a new issues for these or handle them here?

Yes, that'd be very helpful.

btw, I've gone ahead and created #1392 for the 64-bit AppleWin work.

**Comment by tomcw** [comment]
Just an update on commits for this issue: the Feb 8th "good enough" commit (https://github.com/AppleWin/AppleWin/commit/16b2cf329e42d1908f65154d996b938cc4f2553e), got superseded by the Apr 3rd commit (https://github.com/AppleWin/AppleWin/commit/839c493b670eabaa273fa9a901ffd1abfe0ae609), which now supports both a2audit 1.06 (which doesn't support the 1KiB 80-col card, but does some funky stuff which is now properly emulated!) and also supports a2audit 1.07 (which has the fix to support the 1KiB 80-col card).

**Comment by tomcw** [comment]
This issue adds the ability to change the //e's auxiliary slot: `-aux <empty|std80|ext80|rw3>`.
- for `-aux empty`, this includes a new display mode where 80-col will fetch data from the floating bus for the even columns.

Additionally there were extra requirements to include "(build 32-bit)" or "(build 64-bit)" whenever the version is displayed.

Internally, there are new CPU emulation paths to support `-aux empty` and `-aux std80` (and these CPU emulation paths can also be forced by using `-alt-cpu-emu`).
